README.md
# Flutter color picker package

A simple color picker package for your Flutter apps.

## Features

- Pick an RGB color.

## Usage

You can use [KolorPicker] as a normal widget, it is in fact a simple button.
You must give a callback function [onComplete] at least.
The button is by default a square containing an icon [Icons.colorize].
The button is customizable with [icon],[style],[width] and [height] parameters.
You can set an initial color with [initColor], by default it is [Colors.black].
The last picked color is saved and is automatically picked when you reopen the picker.

##### Example

```dart
class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: KolorPicker(onComplete: (c) => print("Picked color $c")),
      ),
    );
  }
}
```

A [BottomSheet] containing the color picker will be opened on [KolorPicker] button click.
The picked color is being returned thank to [onComplete] callback function.

##### Styling picker text

```dart
textTheme: ThemeData.light().textTheme.copyWith(
              bodyText1: TextStyle(
                fontFamily: 'Rubik',
                color: Colors.black,
                fontWeight: FontWeight.w500,
                fontSize: 17,
              ),
              bodyText2: TextStyle(
                fontFamily: 'Rubik',
                color: Colors.blueGrey,
                fontWeight: FontWeight.w700,
                letterSpacing: 2,
                fontSize: 15,
              ),
            ),
```

> bodyText1 is used by slider label ('R', 'G', 'B', 'A')  
> bodyText2 is used by slider value and picked color value


CHANGELOG.md
## 0.0.1 - 2021.06.27

* First version color picker package.
* Simple RGB picker 

## 0.0.2 - 2021.07.16

* Fixed dart file format 


test/kolor_picker_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:kolor_picker/kolor_picker.dart';
import 'package:kolor_picker/src/kolor_slider.dart';
import 'package:kolor_picker/src/picker_modal.dart';
import 'package:kolor_picker/src/rgb_picker.dart';

void main() {
  testWidgets(
    'Kolor picker button test',
    (WidgetTester tester) async {
      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Center(
              child: KolorPicker(onColorPicked: (c) {}),
            ),
          ),
        ),
      );

      print("Test presence button");
      final kolorBtn = find.widgetWithIcon(OutlinedButton, Icons.colorize);
      expect(kolorBtn, findsOneWidget);
    },
  );

  testWidgets(
    'Kolor picker modal test',
    (WidgetTester tester) async {
      const initColor = 0xff030507;
      const modifiedColor = 0xff7f0507;

      await tester.pumpWidget(
        MaterialApp(
          home: Scaffold(
            body: Center(
              child: PickerModal(
                onColorPicked: (c) {},
                initColor: Color(initColor),
              ),
            ),
          ),
        ),
      );

      print("Test presence picker");
      expect(find.byType(RGBPicker), findsOneWidget);

      print("Test presence sliders");
      final sliders = find.byType(KolorSlider);
      expect(sliders, findsNWidgets(4));

      print("Test presence texts");
      final texts = find.byType(Text);
      expect(texts, findsNWidgets(10));

      print("Test value initColor");
      expect(find.text('#${initColor.toRadixString(16)}'), findsOneWidget);

      final redValue = (texts.at(2).evaluate().first.widget as Text).data;
      final greenValue = (texts.at(4).evaluate().first.widget as Text).data;
      final blueValue = (texts.at(6).evaluate().first.widget as Text).data;
      final alphaValue = (texts.at(8).evaluate().first.widget as Text).data;
      print("Test red value");
      expect(redValue, "3");
      print("Test green value");
      expect(greenValue, "5");
      print("Test blue value");
      expect(blueValue, "7");
      print("Test alpha value");
      expect(alphaValue, "255");

      final redSlider = find
          .descendant(
            of: find.byType(KolorSlider),
            matching: find.byType(GestureDetector),
          )
          .at(0);

      await tester.tapAt(tester.getCenter(redSlider).translate(5, 0));
      await tester.pump(Duration(seconds: 1));

      print("Test tap at center of slider changes color");
      final redText =
          (find.byType(Text).at(2).evaluate().first.widget as Text).data;
      expect(redText, "127");
      expect(find.text("#${modifiedColor.toRadixString(16)}"), findsOneWidget);
    },
  );
}


lib/kolor_picker.dart
library kolor_picker;

import 'dart:async';

import 'package:flutter/material.dart';

import 'src/picker_modal.dart';

class KolorPicker extends StatefulWidget {
  /// [KolorPicker] give you the possibility to pick a color RGB values.
  ///
  /// It gives you a button that trigger a [BottomSheet] modal in wich
  /// the user can choose a color. This color can be accessed thank to the
  /// [onColorPicked] callback function.
  ///

  KolorPicker({
    Key? key,
    this.width = 50,
    this.height = 50,
    this.initColor = Colors.black,
    this.icon,
    this.style,
    this.useAppTextTheme = false,
    @required this.onColorPicked,
    this.onComplete,
    this.onError,
  })  : assert(onColorPicked != null),
        super(key: key);

  /// Set the [width] of the button that trigger the picker.
  ///
  final double? width;

  /// Set the [height] of the button that trigger the picker.
  ///
  final double? height;

  /// The initial color that is selected
  ///
  /// Default value [Colors.black]
  ///
  final Color initColor;

  /// Set the [icon] of picker button.
  ///
  /// Default icon is [Icons.colorize].
  ///
  final Icon? icon;

  /// Customize the style of the button that trigger the picker.
  ///
  /// Passing a [style] is not required, a default one is already available.
  ///
  final ButtonStyle? style;

  /// If true the text will use your application custom [textTheme]
  ///
  /// The slider label ('R', 'G', 'B') will use [textTheme.bodyText1]
  ///
  /// The slider value and picked color hex value will use [textTheme.bodyText2]
  ///
  /// By default it is false, it will use a default textTheme.
  ///
  final bool useAppTextTheme;

  /// Callback function to return selected color
  ///
  /// The [onColorPicked] function must not be null
  ///
  /// Paramaters : [Color c]
  /// Return type : void
  ///
  final void Function(Color)? onColorPicked;

  /// Default [onComplete] of [showModalBottomSheet]
  ///
  final Future<void> Function()? onComplete;

  /// Default [onError] of [showModalBottomSheet]
  ///
  final FutureOr<dynamic> Function(Object, StackTrace)? onError;

  @override
  _KolorPickerState createState() => _KolorPickerState();
}

class _KolorPickerState extends State<KolorPicker> {
  final _history = <Color>[];
  Color _pickedColor = Colors.black;

  initState() {
    _pickedColor = widget.initColor;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: widget.width,
      height: widget.height,
      alignment: Alignment.center,
      child: OutlinedButton(
        child: Container(
          width: double.maxFinite,
          height: double.maxFinite,
          child: widget.icon ??
              Icon(
                Icons.colorize,
                size: 20,
              ),
        ),
        style: widget.style ??
            OutlinedButton.styleFrom(
              shape: CircleBorder(),
              side: BorderSide(
                color: _pickedColor,
                width: 3,
                style: BorderStyle.solid,
              ),
              primary: Colors.black,
              onSurface: _pickedColor,
              shadowColor: _pickedColor,
            ),
        onPressed: () {
          showModalBottomSheet(
            context: context,
            backgroundColor: Colors.transparent,
            enableDrag: true,
            builder: (ctx) => Container(
              child: PickerModal(
                history: _history,
                useAppTextTheme: widget.useAppTextTheme,
                onColorPicked: (c) {
                  widget.onColorPicked!(c);
                  setState(() {
                    _pickedColor = c;
                    // * we keep only one occurence of a color
                    if (_history.contains(c)) _history.remove(c);
                    _history.add(c);
                    if (_history.length > 5) _history.removeAt(0);
                  });
                  // * Close the bottom sheet modal
                  Navigator.pop(context);
                },
                initColor: _pickedColor,
              ),
            ),
          )
            ..onError(widget.onError ?? (e, stackTrace) {})
            ..whenComplete(widget.onComplete ?? () {});
        },
      ),
    );
  }
}


lib/src/rgb_picker.dart
import 'package:flutter/material.dart';

import 'kolor_slider.dart';

import 'styles.dart';

class RGBPicker extends StatefulWidget {
  RGBPicker({
    Key? key,
    @required this.onColorChange,
    this.initColor = Colors.black,
    this.useAppTextTheme = false,
  })  : assert(onColorChange != null),
        super(key: key);

  /// The initial color of the picker
  ///
  /// Default value [initColor] = [Colors.black]
  ///
  final Color initColor;

  /// Callback function [onColorChange]
  /// is called when users update one of the sliders value.
  ///
  final void Function(Color)? onColorChange;

  /// If true the text will use your application custom [textTheme]
  ///
  /// The slider label ('R', 'G', 'B') will use [textTheme.bodyText1]
  ///
  /// The slider value and picked color hex value will use [textTheme.bodyText2]
  ///
  /// By default it is false, it will use a default textTheme.
  ///
  final bool useAppTextTheme;

  final _state = _RGBPickerState();

  _RGBPickerState get state => _state;

  @override
  _RGBPickerState createState() => _state;
}

class _RGBPickerState extends State<RGBPicker> {
  int _r = 0, _g = 0, _b = 0, _a = 0;

  @override
  void initState() {
    _a = widget.initColor.alpha;
    _r = widget.initColor.red;
    _g = widget.initColor.green;
    _b = widget.initColor.blue;
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;

    return Container(
      child: SingleChildScrollView(
        physics: BouncingScrollPhysics(),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            // ? Preview selected color
            Container(
              width: 50,
              height: 50,
              margin: EdgeInsets.symmetric(vertical: 10),
              decoration: BoxDecoration(
                border: Border.all(color: Colors.blueGrey[700]!, width: 1),
                color: Color.fromARGB(_a, _r, _g, _b),
                borderRadius: BorderRadius.all(
                  Radius.circular(25),
                ),
              ),
            ),
            Text(
              '#' + Color.fromARGB(_a, _r, _g, _b).value.toRadixString(16),
              style:
                  widget.useAppTextTheme ? textTheme.bodyText2 : valueTextStyle,
            ),
            SizedBox(height: 10),
            KolorSlider(
              initValue: _r,
              useAppTextTheme: widget.useAppTextTheme,
              colors: [
                Color.fromARGB(255, 0, _g, _b),
                Color.fromARGB(255, 255, _g, _b)
              ],
              label: 'R',
              onChanged: (v) {
                _r = v;
                _updateColor();
              },
            ),
            KolorSlider(
              initValue: _g,
              useAppTextTheme: widget.useAppTextTheme,
              colors: [
                Color.fromARGB(255, _r, 0, _b),
                Color.fromARGB(255, _r, 255, _b)
              ],
              label: 'G',
              onChanged: (v) {
                _g = v;
                _updateColor();
              },
            ),
            KolorSlider(
              initValue: _b,
              useAppTextTheme: widget.useAppTextTheme,
              colors: [
                Color.fromARGB(255, _r, _g, 0),
                Color.fromARGB(255, _r, _g, 255)
              ],
              label: 'B',
              onChanged: (v) {
                _b = v;
                _updateColor();
              },
            ),
            KolorSlider(
              initValue: _a,
              useAppTextTheme: widget.useAppTextTheme,
              colors: [
                Color.fromARGB(0, 0, 0, 0),
                Color.fromARGB(255, 0, 0, 0)
              ],
              label: 'A',
              onChanged: (v) {
                _a = v;
                _updateColor();
              },
            )
          ],
        ),
      ),
    );
  }

  _updateColor() {
    setState(() {
      widget.onColorChange!(Color.fromARGB(_a, _r, _g, _b));
    });
  }

  updateColor(Color c) {
    _a = c.alpha;
    _r = c.red;
    _g = c.green;
    _b = c.blue;
    _updateColor();
  }
}


lib/src/kolor_slider.dart
library kolor_slider;

import 'dart:math';

import 'package:flutter/material.dart';

import 'styles.dart';

class KolorSlider extends StatefulWidget {
  /// Custom slider for color picking
  /// with a gradient background to preview color changes
  ///

  KolorSlider({
    Key? key,
    this.width = 250,
    this.height = 13,
    this.initValue = 0,
    this.min = 0,
    this.max = 255,
    this.indicatorColor = Colors.white,
    @required this.colors,
    @required this.label,
    @required this.onChanged,
    this.useAppTextTheme = false,
  })  : assert(onChanged != null),
        assert(colors != null && colors.length >= 2),
        assert(min <= max),
        assert(label != null && label.length == 1),
        super(key: key);

  /// The [width] of the slider
  ///
  /// Default value = 250
  final double width;

  /// The [height] of the slider
  ///
  /// Default value = 16
  ///
  /// The indicator will be sized following this value,
  /// it's size will be : [height] * 1.5
  ///
  final double height;

  /// The initial value of the slider
  ///
  /// Default value [initValue] = 0
  ///
  final int initValue;

  /// Minimum value of the slider
  ///
  /// Default value [min] = 0
  ///
  final int min;

  /// Maximum value of the slider
  ///
  /// Default value [max] = 255
  ///
  final int max;

  /// Colors list that will compose the background gradient
  ///
  /// This parameter can't be null and must be composed of two colors at least.
  ///
  final List<Color>? colors;

  /// Color of the slider indicator
  ///
  /// This value can be null, the default value will be [Colors.white]
  ///
  final Color indicatorColor;

  /// The [label] of the slider reprensent the color (R, G or B)
  ///
  /// Must be one letter
  ///
  final String? label;

  /// Called during a drag when the user is selecting a new value for the slider
  /// by dragging.
  ///
  /// The slider passes the new value to the callback but does not actually
  /// change state until the parent widget rebuilds the slider with the new
  /// value.
  ///
  ///This callback function is required.
  ///
  final void Function(int)? onChanged;

  /// If true the text will use your application custom [textTheme]
  ///
  /// The slider label ('R', 'G', 'B') will use [textTheme.bodyText1]
  ///
  /// The slider value and picked color hex value will use [textTheme.bodyText2]
  ///
  /// By default it is false, it will use a default textTheme.
  ///
  final bool useAppTextTheme;

  @override
  _KolorSliderState createState() => _KolorSliderState();
}

class _KolorSliderState extends State<KolorSlider> {
  double _colorSliderPosition = 0.0;
  int _value = 0;
  double _sliderWidth = 0;

  @override
  Widget build(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;

    const double _margin = 15.0;
    final w = widget.width;
    final h = widget.height;
    final indicatorRadius = (h * 1.5) / 2; //9.75
    final _touchOffset = indicatorRadius / 2 + _margin; //19.875 
    final maxWidth = MediaQuery.of(context).size.width * 4 / 9;

    _sliderWidth = min(w, maxWidth);
    _value = widget.initValue;
    _colorSliderPosition =
        (_value - widget.min) * (_sliderWidth / (widget.max - widget.min));

    return Container(
      margin: EdgeInsets.all(_margin),
      height: h * 2,
      // width: w + indicatorRadius * 2,
      alignment: Alignment.center,
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Expanded(
            child: Container(
              // margin: EdgeInsets.only(right: indicatorRadius * 2),
              alignment: Alignment.center,
              child: Text(
                widget.label!,
                style: widget.useAppTextTheme
                    ? textTheme.bodyText1
                    : labelTextStyle,
              ),
            ),
          ),
          GestureDetector(
            behavior: HitTestBehavior.opaque,
            onHorizontalDragStart: (DragStartDetails details) {
              _colorChangeHandler(details.localPosition.dx - _touchOffset);
            },
            onHorizontalDragUpdate: (DragUpdateDetails details) {
              _colorChangeHandler(details.localPosition.dx - _touchOffset);
            },
            onTapDown: (TapDownDetails details) {
              _colorChangeHandler(details.localPosition.dx - _touchOffset);
            },
            child: Center(
              // ? SLIDER
              child: Container(
                margin: EdgeInsets.symmetric(horizontal: _margin),
                width: _sliderWidth,
                height: h,
                decoration: BoxDecoration(
                  border: Border.all(
                    color: Colors.blueGrey[700]!,
                    width: 1,
                  ),
                  borderRadius: BorderRadius.circular(h / 2),
                  gradient: LinearGradient(colors: widget.colors!),
                ),
                child: CustomPaint(
                  painter: _SliderIndicatorPainter(
                    _colorSliderPosition,
                    indicatorRadius,
                    widget.indicatorColor,
                  ),
                ),
              ),
            ),
          ),
          Expanded(
            child: Container(
              // margin: EdgeInsets.only(left: indicatorRadius * 2),
              alignment: Alignment.center,
              child: Text(
                '$_value',
                style: widget.useAppTextTheme
                    ? textTheme.bodyText2
                    : valueTextStyle,
              ),
            ),
          ),
        ],
      ),
    );
  }

  _colorChangeHandler(double position) {
    //handle out of bounds positions
    if (position > _sliderWidth) {
      position = _sliderWidth;
    }
    if (position < 0) {
      position = 0;
    }
    _colorSliderPosition = position;
    // 0 -> min
    // width -> max
    // pos -> value = (pos*(max-min) / width) + min
    _value =
        (position * (widget.max - widget.min) ~/ _sliderWidth) + widget.min;
    widget.onChanged!(_value);
  }
}

// ? SLIDER INDICATOR
class _SliderIndicatorPainter extends CustomPainter {
  _SliderIndicatorPainter(this.position, this.radius, this.color);

  final double position;
  final double radius;
  final Color color;

  @override
  void paint(Canvas canvas, Size size) {
    // * Background border
    double y = size.height / 2;
    canvas.drawCircle(
      Offset(position, y),
      radius,
      Paint()..color = Colors.blueGrey[700]!,
    );
    // * Indicator front circle
    canvas.drawCircle(
      Offset(position, y),
      radius * 0.9,
      Paint()..color = this.color,
    );
  }

  @override
  bool shouldRepaint(_SliderIndicatorPainter old) {
    return true;
  }
}

// Some parts of this widget code is from : https://medium.com/@mhstoller.it/how-i-made-a-custom-color-picker-slider-using-flutter-and-dart-e2350ec693a1


lib/src/picker_modal.dart
library picker_modal;

import 'package:flutter/material.dart';

import 'styles.dart';
import 'rgb_picker.dart';

class PickerModal extends StatelessWidget {
  const PickerModal({
    Key? key,
    this.initColor = Colors.black,
    this.useAppTextTheme = false,
    @required this.onColorPicked,
    this.history,
  })  : assert(onColorPicked != null),
        super(key: key);

  /// The initial color of the picker
  ///
  /// Default value [initColor] = [Colors.black]
  ///
  final Color initColor;

  /// Callback function [onColorPicked]
  /// is called when users click on the
  /// button to confirm their choice.
  ///
  final void Function(Color)? onColorPicked;

  /// If true the text will use your application custom [textTheme]
  ///
  /// The slider label ('R', 'G', 'B') will use [textTheme.bodyText1]
  ///
  /// The slider value and picked color hex value will use [textTheme.bodyText2]
  ///
  /// By default it is false, it will use a default textTheme.
  ///
  final bool useAppTextTheme;

  /// Last 5 colors picked.
  ///
  final List<Color>? history;

  @override
  Widget build(BuildContext context) {
    final textTheme = Theme.of(context).textTheme;
    Color _currentColor = initColor;
    final bottomMargin = MediaQuery.of(context).viewPadding.bottom;

    final picker = RGBPicker(
      useAppTextTheme: useAppTextTheme,
      initColor: initColor,
      onColorChange: (color) {
        _currentColor = color;
      },
    );

    return Container(
      height: 900,
      margin: EdgeInsets.only(
        left: 10,
        right: 10,
        top: 0,
        bottom: bottomMargin,
      ),
      padding: EdgeInsets.all(10),
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.all(Radius.circular(15.0)),
      ),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.start,
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          Container(
            width: 50,
            height: 5,
            margin: EdgeInsets.only(bottom: 15),
            decoration: BoxDecoration(
              color: Colors.black38,
              borderRadius: BorderRadius.all(
                Radius.circular(10),
              ),
            ),
          ),
          Expanded(
            child: picker,
          ),
          SizedBox(height: 20),
          Row(
            children: [
              if (history != null)
                Expanded(
                  child: Row(
                    children: history!.reversed
                        .map((c) => _historyButton(c, picker))
                        .toList(),
                  ),
                ),
              Container(
                alignment: Alignment.centerRight,
                padding: EdgeInsets.symmetric(horizontal: 10),
                child: OutlinedButton(
                  onPressed: () => onColorPicked!(_currentColor),
                  child: Text(
                    'Ok',
                    style:
                        useAppTextTheme ? textTheme.bodyText1 : labelTextStyle,
                  ),
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _historyButton(Color c, RGBPicker picker) {
    return Container(
      width: 30,
      height: 30,
      margin: EdgeInsets.all(3),
      child: OutlinedButton(
        onPressed: () {
          picker.state.updateColor(c);
        },
        style: OutlinedButton.styleFrom(
          backgroundColor: c,
        ),
        child: Container(),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';

import 'package:kolor_picker/kolor_picker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        textTheme: ThemeData.light().textTheme.copyWith(
              bodyText1: TextStyle(
                fontFamily: 'Rubik',
                color: Colors.black,
                fontWeight: FontWeight.w500,
                fontSize: 17,
              ),
              bodyText2: TextStyle(
                fontFamily: 'Rubik',
                color: Colors.blueGrey,
                fontWeight: FontWeight.w700,
                letterSpacing: 2,
                fontSize: 15,
              ),
            ),
      ),
      home: MyHomePage(title: 'kolor_picker example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: KolorPicker(
          onColorPicked: (c) => print("Picked color $c"),
          // useAppTextTheme: true,
        ),
      ),
    );
  }
}
